torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -839,6 +852,17 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
   bool storeError_{false};
 
+  // The lock which protects the write/read of
+  // extendedTimeout_/inflightTimeoutExt_.
+  std::mutex mtxTimeoutExtension_;
+
+  // The extended timeout on top of existing timeout for works issued before
+  // first work finishes.
+  std::chrono::milliseconds extendedTimeout_ = std::chrono::milliseconds(0);
+
+  // The timeout extension which has been already applied to work.
+  std::chrono::milliseconds inflightTimeoutExt_ = std::chrono::milliseconds(0);
```

Comment:
It's weird that the first variable is named extendedTimeout but the second is TimeoutExt. Pick... a consistent ordering?

Reply:
Sure. Originally, I want to just call it `inflightTimeout_` but here it only means the extended part.. OK. I can make it consistent.

Reply:
We'd better minimize the states and so many different definition of timeouts make things hard to read. Is this additional timeout necessary? extendedTimeout would be set by user, and reset to 0 when the first work completes. When checkout timeout and extendedTimeout_ > 0,  you can pass the customized timeout to ProcessGroupNCCL::WorkNCCL::checkTimeout().

Reply:
This one is needed, because you aren't supposed to necessarily reset to zero when the first work completes. The interleaved case can be seen in the spreadsheet in PR desc.